src/main/java/org/littleshoot/proxy/impl/ProxyConnection.java

```diff
@@ -789,11 +789,15 @@ public void write(ChannelHandlerContext ctx,
         }
 
         @Override
-        public void channelRead(ChannelHandlerContext ctx, Object msg)
-            throws Exception {
+        public void channelRead(ChannelHandlerContext ctx, Object msg) {
             try {
                 proxyServer.getGlobalStateHandler().restoreFromChannel(clientToProxyConnection.channel);
                 super.channelRead(ctx, msg);
+            } catch (Throwable e) {
+                if (msg instanceof ReferenceCounted) {
+                    ((ReferenceCounted)msg).release();
+                }
+                throw new RuntimeException("Failed to execute inbound global state handler", e);
```

Comment:
@viacheslav-fomin-main let's present the error handling types intact for now. i don't want to introduce too many changes.

Reply:
@viacheslav-fomin-main 
```
@Override
    public final void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
            throws Exception {
        exceptionCaught(cause);
    }
```

it could be simpler to move this logic to 
org.littleshoot.proxy.impl.ProxyConnection#exceptionCaught(io.netty.channel.ChannelHandlerContext, java.lang.Throwable)